mernarezk/assignment3/Dictionary.java

```diff
@@ -0,0 +1,52 @@
+package Assignment3;
+
+import java.util.HashSet;
+import java.util.Set;
+/**
+ * Simple dictionary class that has two attributes, a set of words which are assumed to be only
+ * lower case english letters and prefixes set which is a set that has all possible prefixes of all words.
+ */
+public class Dictionary {
+	private Set<String> words;
+	private Set<String> prefixes;
+
+	public Dictionary() {
+		words = new HashSet<>();
+		prefixes = new HashSet<>();
+	}
+	/**
+	 * Each time we add a new word the dictionary, we compute all possible prefixes and add them to a 
+	 * set of prefixes to eliminate duplicates generated from words like cat,cat,cape .. etc.
+	 * @param s new word to be added to the dictionary
+	 */
+	public void addWord(String s) {
+		generatePrefixes(s);
+		words.add(s);
+	}
+
+	private Set<String> generatePrefixes(String s) {
+		StringBuilder str = new StringBuilder("");
+		for (int i = 0; i < s.length(); i++) {
+			str.append(s.charAt(i));
+			prefixes.add(str.toString());

```

Comment:
It could also be:

```
for (int i = 1; i <= s.length(); i++) {
    prefixes.add(str.substring(0, i));
}
```

Reply:
oh yes but I prefer using stringbuilder because it takes less time to append to a string O(n) while substring takes O(n^2) since strings are immutable in java.